File: 剑指Offer/68.树中两个节点的最低公共祖先2.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def lowestCommonAncestor(self, root: TreeNode, p: TreeNode, q: TreeNode) -> TreeNode:
        if root is None:
            return None
        
        if p.val == root.val or q.val == root.val:
            return root
        
        right = self.lowestCommonAncestor(root.right, p, q)
        left = self.lowestCommonAncestor(root.left, p, q)

        if left is not None and right is not None:
            return root
        else:
            return right if right is not None else left

File: 剑指Offer/68.树中两个节点的最低公共祖先2.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    TreeNode* lowestCommonAncestor(TreeNode* root, TreeNode* p, TreeNode* q) {
        if (root == NULL) {
            return NULL;
        }

        if (p->val == root->val || q->val == root->val) {
            return root;
        }

        TreeNode *right, *left;
        right = lowestCommonAncestor(root->right, p, q);
        left = lowestCommonAncestor(root->left, p, q);
        
        if (right != NULL && left != NULL) {
            return root;
        }
        else return right == NULL ? left : right;
    }
};